src/aklite_client_ext.cc

```diff
@@ -282,3 +283,45 @@ bool AkliteClientExt::RebootIfRequired() {
 
   return false;
 }
+
+InstallResult AkliteClientExt::Rollback(const LocalUpdateSource* local_update_source) {
+  auto current = GetCurrent();
+  auto pending_target = GetPendingTarget();
+  auto installation_in_progress = !pending_target.IsUnknown();
+  auto bad_target = installation_in_progress ? pending_target : current;
+
+  LOG_DEBUG << "User initiated rollback. Current Target is " << current.Name();
+  if (installation_in_progress) {
+    LOG_DEBUG << "There is an installation in progress of Target " << pending_target.Name();
+  }
+  auto storage = INvStorage::newStorage(client_->config.storage, false);
+  LOG_INFO << "Marking target " << bad_target.Name() << " as failed";
+  storage->saveInstalledVersion("", Target::fromTufTarget(bad_target), InstalledVersionUpdateMode::kBadTarget);
+
+  // Get rollback target
+  auto rollback_target = GetRollbackTarget(true, installation_in_progress);
+  if (rollback_target.IsUnknown()) {
+    LOG_ERROR << "Failed to find Target to rollback to";
+    return {InstallResult::Status::Failed};
+  }
+
+  if (installation_in_progress) {
+    // Previous installation was not finalized
+    LOG_INFO << "Creating new installation log entry for " << rollback_target.Name() << ", as we try to rollback to it";
+    storage->saveInstalledVersion("", Target::fromTufTarget(rollback_target), InstalledVersionUpdateMode::kNone);
+  }
+
+  std::string reason = "User initiated rollback. Marked " + bad_target.Name() +
+                       " as a failing target, and rolling back to " + rollback_target.Name();
+  LOG_INFO << reason;
+  // First, try to rollback without download
+  auto ret = PullAndInstall(rollback_target, reason, "", InstallMode::All, local_update_source, false, true,
+                            !installation_in_progress);
+  if (ret.status == InstallResult::Status::DownloadFailed) {
+    LOG_INFO
+        << "Available data for rollback not fully available. Trying to download data before trying to install again";
```

Comment:
Maybe `Not all rollback Target data are available, downloading them before the next install attempt...`?